File: src/main/java/se/uglisch/resource/Resource.java
package se.uglisch.resource;

import java.io.InputStream;

import javax.xml.transform.stream.StreamSource;

public class Resource extends StreamSource {

	private final String value;

	private Resource(final String value) {
		this.value = value;
		setInputStream(toStream());
	}

	public static Resource apply(final String value) {
		return new Resource(value);
	}

	private InputStream toStream() {
		final InputStream stream = getClass().getResourceAsStream(value);
		if (stream == null) {
			throw new ResourceNotFoundException(value);
		}
		return stream;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((value == null) ? 0 : value.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Resource other = (Resource) obj;
		if (value == null) {
			if (other.value != null)
				return false;
		} else if (!value.equals(other.value))
			return false;
		return true;
	}

}

File: src/main/scala/se/uglisch/resource/Resource.scala
package se.uglisch.resource

import java.io.InputStream
import java.io.Reader
import java.io.StringReader
import java.io.StringWriter
import org.apache.commons.io.IOUtils
import javax.xml.transform.Source
import javax.xml.transform.stream.StreamSource
import java.io.File
import java.net.URL

case class Resource(path: String) {

  def asStream: Option[InputStream] =
    Option(getClass.getResourceAsStream(path))

  def asSource: Option[Source] = asStream match {
    case Some(stream) => Option(new ResourceSource(stream))
    case None => None
  }

  def asUrl: Option[URL] =
    Option(getClass.getResource(path))

  def asFile: Option[File] = asUrl match {
    case Some(url) => Option(new File(url.getFile))
    case None => None
  }
}

class ResourceSource(stream: InputStream) extends StreamSource {

  val data = {
    val writer = new StringWriter
    IOUtils.copy(stream, writer)
    writer.toString
  }

  override def getReader: Reader =
    new StringReader(data)

}

